src/core/net/nat64_translator.cpp

```diff
@@ -345,61 +378,139 @@ uint16_t Translator::ReleaseExpiredMappings(void)
 
     return ReleaseMappings(idleMappings);
 }
+#if OPENTHREAD_CONFIG_NAT64_PORT_TRANSLATION_ENABLE
+uint16_t Translator::AllocateSourcePort(uint16_t aSrcPort)
+{
+    // The translated port is randomly allocated from the range of dynamic or private ports
+    // to prevent strange situations where the translated source port might be a well-known port
+    uint16_t retPort;
+
+    while (true)
+    {
+        retPort = Random::NonCrypto::GetUint16InRange(kTranslationPortRangeStart, kTranslationPortRangeEnd);
+        // The NAT64 SHOULD preserve the port parity (odd/even), as per Section 4.2.2 of [RFC4787]).
+        // Determine if original and allocated port have different parity
+        if (((aSrcPort ^ retPort) & 1) == 1)
+        {
+            retPort++;
+        }
+        if (!mActiveAddressMappings.ContainsMatching(retPort))
+        {
+            break;
+        }
+    }
+
+    return retPort;
+}
+#endif
 
-Translator::AddressMapping *Translator::AllocateMapping(const Ip6::Address &aIp6Addr)
+Translator::AddressMapping *Translator::AllocateMapping(const Ip6::Address &aIp6Addr,
+                                                        const uint16_t      aSrcPortOrId,
+                                                        const uint8_t       aProtocol)
```

Comment:
We can consider passing `Ip6::Headers` or `Ip4::Headers` directly to `AllocateMapping()` as well?

Reply:
Yes, looks good.